source/agora/node/TransactionRelayer.d

```diff
@@ -0,0 +1,462 @@
+
+    private string addTransactionImp (in Transaction tx) @safe nothrow
+    {
+        auto tx_hash = tx.hashFull();
+        if (tx_hash in this.tx_hashes)
+            return TransactionRelayError.TxAlreadyAdded;
+
+        Amount fee;
+        if (auto fee_err = this.getAdjustedTXFee(tx, fee))
+            return fee_err;
+        else if (!fee.isValid())
+            return TransactionRelayError.TxFeeInvalid;
+
+        if (fee < this.config.node.relay_tx_min_fee)
+            return TransactionRelayError.TxFeeTooLow;
+
+        this.txs.insert(TxHolder(fee, tx_hash,
+            this.clock.networkTime() + this.config.node.relay_tx_cache_exp.total!"seconds"));
+        this.tx_hashes.put(tx_hash);
+
+        return null;
+    }
+
+    /***************************************************************************
+
+        Adds a transaction to the relay queue. Unlike addTransactionImp, this
+        method can have stronger side-effects like logging or even triggering
+        transaction relay. This function should only be called, after
+        the transaction was succesfully added to the transaction pool.
+
+        Params:
+            tx = transaction to be added to the relay queue
+
+        Returns:
+            null, if there was no error, otherwise string detailing the error
+
+    ***************************************************************************/
+
+    public override string addTransaction(in Transaction tx)  @safe nothrow
+    {
+        string res = this.addTransactionImp(tx);
+        if (res)
+            this.log.trace(res);
+        else if (this.config.node.relay_tx_interval == 0.seconds)
+            this.relayTransactions();
+        return res;
+    }
+
+    ///
+    public static enum TransactionRelayError : string
+    {
+        TxAlreadyAdded = "Transaction has already been added to the relay queue",
+        TxFeeInvalid = "Transaction fee is invalid",
+        TxFeeTooLow = "Transaction fee is too low"
+    }
+
+    ///
+    private static struct TxHolder
+    {
+        /// Transaction fee.
+        public Amount fee;
+
+        /// Hash of the transaction.
+        public Hash hash;
+
+        /// The time when this entry should be removed from cache.
+        public TimePoint expiry;
+
+        ///
+        public static bool cmpFeeLess (in TxHolder tx_holder1, in TxHolder tx_holder2) @safe nothrow
+        {
+            return tx_holder1.fee > tx_holder2.fee;
+        }
+
+        ///
+        public void toString (scope SinkT dg) const @trusted
+        {
+            import std.format;
+
+            // Workaround for missing @safe on the delegate passed to RedBlackTree.toString().
+            auto safe_dg = cast(SafeSinkT) dg;
+
+            formattedWrite(safe_dg, "{ amount: ");
+            this.fee.toString(safe_dg);
+            formattedWrite(safe_dg, ", hash: ");
+            this.hash.toString(safe_dg);
+            formattedWrite(safe_dg, ", expiry: %d }", this.expiry);
```

Comment:
You could just merge it all into one statement, `formattedWrite` recognizes `sink`